app/AvailablePlugin/PrivacyIdeaAuthenticator/Controller/PaperTokensController.php

```diff
@@ -0,0 +1,198 @@
+ *
+ * Portions licensed to the University Corporation for Advanced Internet
+ * Development, Inc. ("UCAID") under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * UCAID licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ * @link          http://www.internet2.edu/comanage COmanage Project
+ * @package       registry-plugin
+ * @since         COmanage Registry v4.3.0
+ * @license       Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ */
+
+App::uses("SAMController", "Controller");
+
+class PaperTokensController extends SAMController {
+  // Class name, used by Cake
+  public $name = "PaperTokens";
+
+  public $requires_person = true;
+
+  public $view_contains = array(
+    'PrivacyIdeaAuthenticator' => array('Server')
+  );
+
+  /**
+   * Add a Standard Object.
+   *
+   * @since  COmanage Registry v4.3.0
+   */
+
+  public function generate() {
+    if($this->request->is('get')) {
+      parent::add();
+
+      //$this->set('title_for_layout', 'Generated '.$this->viewVars['vv_authenticator']['Authenticator']['description']);
+      $this->set('title_for_layout', 'Generated Backup Codes');
+
+      if(!empty($this->request->params['named']['onFinish'])) {
+        $this->set('vv_on_finish_url', $this->request->params['named']['onFinish']);
+      }
+
+      try {
+        $tokenInfo = $this->PrivacyIdea->createToken($this->viewVars['vv_authenticator']['PrivacyIdeaAuthenticator'],
+                                                     $this->viewVars['vv_co_person']['CoPerson']['id']);
+
+        $this->set('vv_token_info', $tokenInfo);
+
+        $newdata = array(
+          'PaperToken' => array(
+            'co_person_id' => $this->viewVars['vv_co_person']['CoPerson']['id'],
+            'serial' => $tokenInfo['serial']
+          )
+        );
+
+        $this->generateHistory('generate', $newdata, array());
+
+        if(!empty($tokenInfo['otps'])) {
+          $this->set('vv_otps', (array)$tokenInfo['otps']);
+          debug($vv_otps);
+        }
+      }
+      catch(Exception $e) {
+        $this->Flash->set($e->getMessage(), array('key' => 'error'));
+      }
+    }
+  }
+
+  /**
+   * Callback before other controller methods are invoked or views are rendered.
+   *
+   * @since  COmanage Registry v4.3.0
+   */
+
+  public function beforeFilter() {
```

Comment:
Is there a reason why not to use the recommended CAKEPHP approach? if you wish to use a Model in your controller: https://book.cakephp.org/2/en/controllers.html#components-helpers-and-uses

e.g.
```
  public $uses = array(
    'EmailAddressWidget.CoEmailAddressWidget',
    'EmailAddress',
    'CO',
    'CoMessageTemplate'
  );
  ```
when using `$uses` the first element of the array should always be this controller's model.